File: evo-ai-frontend-community/src/services/integrations/canvaService.ts
import api from '@/services/core/api';
import type {
  CanvaConfig,
  CanvaOAuthResponse,
  CanvaConnectionResponse,
  DiscoverToolsResponse
} from '@/types/integrations';

const CanvaService = {
  /**
   * Generate Canva OAuth authorization URL
   */
  async generateAuthorization(agentId: string): Promise<CanvaOAuthResponse> {
    try {
      const { data } = await api.post(
        `/agents/${agentId}/integrations/canva/authorization`
      );
      return data;
    } catch (error) {
      console.error('CanvaService.generateAuthorization error:', error);
      throw error;
    }
  },

  /**
   * Complete Canva OAuth flow
   */
  async completeAuthorization(
    agentId: string,
    code: string,
    state: string
  ): Promise<CanvaConnectionResponse> {
    try {
      const { data } = await api.post(
        `/agents/${agentId}/integrations/canva/callback`,
        {
          code,
          state,
        }
      );
      return data;
    } catch (error) {
      console.error('CanvaService.completeAuthorization error:', error);
      throw error;
    }
  },

  /**
   * Get Canva integration configuration
   */
  async getConfiguration(agentId: string): Promise<CanvaConfig | null> {
    try {
      const { data } = await api.get(
        `/agents/${agentId}/integrations/canva`
      );
      return data.config || null;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error) {
        const httpError = error as { response?: { status?: number } };
        if (httpError.response?.status === 404) {
          return null;
        }
      }
      console.error('CanvaService.getConfiguration error:', error);
      throw error;
    }
  },

  /**
   * Save Canva integration configuration
   */
  async saveConfiguration(
    agentId: string,
    config: Partial<CanvaConfig>
  ): Promise<{ success: boolean }> {
    try {
      const { data } = await api.put(
        `/agents/${agentId}/integrations/canva`,
        { config }
      );
      return data;
    } catch (error) {
      console.error('CanvaService.saveConfiguration error:', error);
      throw error;
    }
  },

  /**
   * Disconnect Canva integration
   */
  async disconnect(agentId: string): Promise<{ success: boolean }> {
    try {
      const { data } = await api.delete(
        `/agents/${agentId}/integrations/canva`
      );
      return data;
    } catch (error) {
      console.error('CanvaService.disconnect error:', error);
      throw error;
    }
  },

  /**
   * Discover available MCP tools from Canva
   * Uses backend endpoint that handles access_token internally
   */
  async discoverTools(agentId: string): Promise<DiscoverToolsResponse> {
    try {
      const { data } = await api.get(
        `/agents/${agentId}/integrations/canva/discover-tools`
      );
      return data;
    } catch (error) {
      console.error('CanvaService.discoverTools error:', error);
      throw error;
    }
  },
};

export default CanvaService;

File: evo-ai-processor-community/src/services/canva_service.py
"""
Canva MCP OAuth and API integration service.
This service uses the MCP OAuth Authorization Server discovery pattern:
1. Discover OAuth requirements from Canva MCP server
2. Perform OAuth flow using discovered authorization server
3. Store tokens for use in MCP server headers
This follows the standard MCP OAuth pattern for Canva MCP.
"""
import logging
from typing import Optional, Dict, Any
import httpx
from src.services.mcp_oauth_service import MCPOAuthService
logger = logging.getLogger(__name__)
class CanvaService(MCPOAuthService):
    """Service for Canva MCP OAuth and API operations."""
    # Default Canva MCP URL
    DEFAULT_MCP_URL = "https://mcp.canva.com"
    def __init__(
        self,
        redirect_uri: str,
        core_service_url: str,
        user_token: str,
        client_id: Optional[str] = None,
        client_secret: Optional[str] = None,
        mcp_url: Optional[str] = None
    ):
        """
        Initialize Canva MCP service.
        Args:
            redirect_uri: OAuth redirect URI
            core_service_url: Base URL for evo-ai-core-service API
            user_token: User authentication token for API calls
            client_id: Canva OAuth client ID (optional, can be obtained via dynamic registration)
            client_secret: Canva OAuth client secret (optional, can be obtained via dynamic registration)
            mcp_url: Optional custom MCP URL (defaults to Canva MCP)
        """
        effective_mcp_url = mcp_url or self.DEFAULT_MCP_URL
        super().__init__(
            mcp_url=effective_mcp_url,
            redirect_uri=redirect_uri,
            core_service_url=core_service_url,
            user_token=user_token,
            provider_name="canva",
            client_id=client_id,
            client_secret=client_secret
        )
        logger.info(f"CanvaService initialized with MCP URL: {effective_mcp_url}")
    async def complete_authorization(
        self,
        agent_id: str,
        code: str,
        state: str,
        db: Optional[Any] = None
    ) -> Dict[str, Any]:
        """
        Complete OAuth authorization flow and store tokens.
        Overrides parent method to add Canva-specific user info retrieval.
        Args:
            agent_id: Agent ID
            code: Authorization code from OAuth callback
            state: State parameter from OAuth callback
            db: Optional database session for direct DB access
        Returns:
            Dictionary with success status, username, and user info
        """
        # Call parent method to handle OAuth flow
        result = await super().complete_authorization(
            agent_id=agent_id,
            code=code,
            state=state,
            db=db
        )
        # If successful, get Canva user info
        if result.get("success") and not result.get("username"):
            try:
                access_token = None
                # Get access token from stored credentials
                credentials = await self._load_credentials(agent_id)
                if credentials:
                    access_token = credentials.get("access_token")
                if access_token:
                    # Get Canva user info
                    async with httpx.AsyncClient() as client:
                        user_response = await client.get(
                            f"{self.mcp_url}/user",
                            headers={
                                "Authorization": f"Bearer {access_token}",
                                "Content-Type": "application/json"
                            },
                            timeout=30.0
                        )
                        user_response.raise_for_status()
                        user_data = user_response.json()
                        result["username"] = user_data.get("name") or user_data.get("email") or user_data.get("id")
                        result["email"] = user_data.get("email")
                        result["user_data"] = user_data
                        # Update stored credentials with user info
                        await self._store_credentials(
                            agent_id=agent_id,
                            mcp_url=self.mcp_url,
                            access_token=access_token,
                            username=result["username"],
                            email=result["email"],
                            user_data=user_data,
                            oauth_metadata=self._oauth_metadata or {}
                        )
            except Exception as e:
                logger.warning(f"Could not fetch Canva user info: {e}")
                # Don't fail the authorization if user info fetch fails
        return result
    async def generate_authorization_url(
        self,
        agent_id: str,
        scopes: Optional[list] = None,
    ) -> str:
        """
        Generate OAuth 2.0 authorization URL for Canva MCP.
        First discovers OAuth requirements from the MCP server, then generates
        the authorization URL using the discovered authorization server and scopes.
        Args:
            agent_id: Agent ID
            scopes: Optional list of scopes to request (defaults to all supported scopes)
        Returns:
            Authorization URL for user to visit
        """
        # Discover OAuth requirements if not already discovered
        if not self._oauth_metadata:
            logger.info("Discovering OAuth requirements from Canva MCP server...")
            await self.discover_oauth_requirements()
        # Generate authorization URL using parent method
        return await super().generate_authorization_url(
            agent_id=agent_id,
            scopes=scopes
        )
